File: src/tmp/Python/array/find_all_duplicates_in_an_array.py
from typing import List

def findDuplicates(nums: List[int]) -> List[int]:
    n = len(nums)
    result = []
    for num in nums:
        num = (num - 1) % n
        if nums[num] > n:
            result.append(num + 1)
        nums[num] += n
    
    return result

if __name__ == "__main__":
    print(findDuplicates([2,2]))

File: src/tmp/Java/src/com/dbc/algorithm/Array/FindAllDuplicatesInAnArray.java
package com.dbc.algorithm.Array;

import java.util.ArrayList;
import java.util.List;

public class FindAllDuplicatesInAnArray {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> result = new ArrayList<>();
        for (int num : nums) {
            num = (num - 1) % nums.length;
            if (nums[num] > nums.length) {
                result.add(num + 1);
            }
            nums[num] += nums.length;
        }
        return result;
    }
}
